cli/src/daemon.rs

```diff
@@ -129,104 +192,126 @@ struct Snapshot {
-                // hammer the RPC again straight away. Even better would be to do
-                // exponential backoff with jitter, but let's not go there right now.
-                do_wait = true;
+    /// Mutex where we publish the latest snapshot for use by the webserver.
+    snapshot_mutex: Arc<SnapshotMutex>,
+}
 
-                None
-            }
+impl<'a, 'b> Daemon<'a, 'b> {
+    pub fn new(config: &'a mut SnapshotClientConfig<'b>, opts: &'a RunMaintainerOpts) -> Self {
+        let metrics = MaintenanceMetrics {
+            polls: 0,
+            errors: 0,
+            transactions_stake_deposit: 0,
+            transactions_update_exchange_rate: 0,
+            transactions_withdraw_inactive_stake: 0,
+            transactions_collect_validator_fee: 0,
+            transactions_merge_stake: 0,
+            transactions_claim_validator_fee: 0,
+            transactions_unstake_from_inactive_validator: 0,
+            transactions_remove_validator: 0,
         };
+        Daemon {
+            config,
+            opts,
+            rng: rand::thread_rng(),
+            last_read_success: Instant::now(),
+            metrics,
+            snapshot_mutex: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    /// Publish a new snapshot that from now on will be served by the http server.
+    fn publish_snapshot(&mut self, solido: Option<SolidoState>) {
+        if solido.is_some() {
+            self.last_read_success = Instant::now();
+        }
 
-        // Publish the new state and metrics, so the webserver can serve them.
         let snapshot = Snapshot {
-            metrics: metrics.clone(),
-            solido: state,
+            metrics: self.metrics.clone(),
+            solido,
         };
-        snapshot_mutex.lock().unwrap().replace(Arc::new(snapshot));
-
-        if do_wait {
-            // Sleep a random time, to avoid a thundering herd problem, in case
-            // multiple maintainer bots happened to run in sync. They would all
-            // try to create the same transaction, and only one would pass.
-            let max_poll_interval = Duration::from_secs(*opts.max_poll_interval_seconds());
-            let sleep_time = rng.gen_range(Duration::from_secs(0)..max_poll_interval);
-            println!("Sleeping {:?} until next iteration ...", sleep_time);
-            std::thread::sleep(sleep_time);
+        self.snapshot_mutex
+            .lock()
+            .unwrap()
+            .replace(Arc::new(snapshot));
+    }
+
+    /// Sleep with exponential backoff and jitter.
+    fn sleep_after_error(&mut self) {
+        // For the sleep time we use exponential backoff with jitter [1]. By taking
+        // the time since the last success as the target sleep time, we get
+        // exponential backoff. We clamp this to ensure we don't wait indefinitely.
+        // 1: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
```

Comment:
The maintainer just got fancier now ^^, but where's the exponential backoff part of it?
Great!

Reply:
The sleep time is the time since the last successful call, see below. This makes the time between calls grow exponentially, as it roughly doubles every time: the time you sleep this time is the sum of all previous sleeps.